### PR TITLE
8318629: G1: Refine code a bit in G1RemSetTrackingPolicy::update_at_allocate

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you help to review this simple refactoring?
Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318629](https://bugs.openjdk.org/browse/JDK-8318629): G1: Refine code a bit in G1RemSetTrackingPolicy::update_at_allocate (**Enhancement** - P4)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [9c4149e1](https://git.openjdk.org/jdk/pull/16300/files/9c4149e15817295b765e06e70d02634e50826e2c)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [9c4149e1](https://git.openjdk.org/jdk/pull/16300/files/9c4149e15817295b765e06e70d02634e50826e2c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16300/head:pull/16300` \
`$ git checkout pull/16300`

Update a local copy of the PR: \
`$ git checkout pull/16300` \
`$ git pull https://git.openjdk.org/jdk.git pull/16300/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16300`

View PR using the GUI difftool: \
`$ git pr show -t 16300`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16300.diff">https://git.openjdk.org/jdk/pull/16300.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16300#issuecomment-1773768276)

### CODE REVIEW

**Comment by Hamlin-Li** [comment]
@albertnetymk Thanks for the suggestion, I think it's better.

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1RemSetTrackingPolicy.cpp [hunk 1]
```diff
@@ -38,18 +38,16 @@ bool G1RemSetTrackingPolicy::needs_scan_for_rebuild(HeapRegion* r) const {
...
+        "Region %u with unexpected heap region type %s", r->hrm_index(), r->get_type_str());
+  if (r->is_old()) {
     // By default, do not create remembered set for new old regions.
     r->rem_set()->set_state_untracked();
-  } else {
-    guarantee(false, "Unhandled region %u with heap region type %s", r->hrm_index(), r->get_type_str());
+    return;
   }
+  // Always collect remembered set for young regions.
+  // Collect remembered sets for humongous regions by default to allow eager reclaim.
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1RemSetTrackingPolicy.cpp:49]:
```suggestion
  // Always collect remembered set for young regions and for humongous regions. This allows eager reclaim for the latter.
```

**Review by tschatzl — COMMENTED** [review]
Fwiw, the original intent of the existing code structure has been to be _very_ obvious and clear with what happens to which region to avoid mistakes.

Imo this change makes the code more clear, but if more people think it does not matter I'm not opposed to it.

**Review by tschatzl — APPROVED** [review]
Then again, it's not that important, but please fix the comment first to not sound that disjointed.

**Comment by Hamlin-Li** [comment]
Thanks Thomas for the suggestions.

**Comment by Hamlin-Li** [comment]
@albertnetymk @tschatzl Thanks for your reviewing.

**Pull request closed without merging**